matchers_test.go

```diff
@@ -315,3 +338,177 @@ func TestConservativeMatcher(t *testing.T) {
 		})
 	}
 }
+
+func TestDefinitelySafe(t *testing.T) {
+	type TestCase struct {
+		value string
+		want  string
+	}
+
+	testCases := map[string]TestCase{
+		"literals, boolean, true": {
+			value: `echo ${{ true }}`,
+			want:  `echo `,
+		},
+		"literals, boolean, false": {
+			value: `echo ${{ false }}`,
+			want:  `echo `,
+		},
+		"literals, null": {
+			value: `echo ${{ null }}`,
+			want:  `echo `,
+		},
+		"literals, number, integer": {
+			value: `echo ${{ 42 }}`,
+			want:  `echo `,
+		},
+		"literals, number, negative integer": {
+			value: `echo ${{ -36 }}`,
+			want:  `echo `,
+		},
+		"literals, number, float": {
+			value: `echo ${{ 3.14 }}`,
+			want:  `echo `,
+		},
+		"literals, number, hexadecimal (uppercase)": {
+			value: `echo ${{ 0x2A }}`,
+			want:  `echo `,
+		},
+		"literals, number, hexadecimal (lowercase)": {
+			value: `echo ${{ 0x2a }}`,
+			want:  `echo `,
+		},
+		"literals, number, hexadecimal (mixed case)": {
+			value: `echo ${{ 0xDeAdBeEf }}`,
+			want:  `echo `,
+		},
+		"literals, number, scientific notation, positive, small": {
+			value: `echo ${{ 2.99e-2 }}`,
+			want:  `echo `,
+		},
+		"literals, number, scientific notation, positive, big": {
+			value: `echo ${{ 2.99e2 }}`,
+			want:  `echo `,
+		},
+		"literals, number, scientific notation, negative, small": {
+			value: `echo ${{ -2.99e-2 }}`,
+			want:  `echo `,
+		},
+		"literals, number, scientific notation, negative, big": {
+			value: `echo ${{ -2.99e2 }}`,
+			want:  `echo `,
+		},
+		"literals, string": {
+			value: `echo ${{ 'Hello world' }}`,
+			want:  `echo `,
+		},
+		"literals, multiple": {
+			value: `echo ${{ true || 42 }}`,
+			want:  `echo ${{  ||  }}`,
```

Comment:
I suppose a more elegant implementation would also resolve unnecessary [operators](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#operators) like this. However, I believe for the currently intended use of this function, that does not add any functional value.

Reply:
Glad to have left this overnight, because this should definitely be handled. Otherwise the non`-conservative` mode will still complain about this expression.

Moreover, in working on this I realized that there's more cleanup one can do. For example, if all arguments are removed from a function (because they're all literals, for some reason) than that function can always be removed - further pruning out safe expressions.